nibabel/streamlines/tractogram.py

```diff
@@ -0,0 +1,543 @@
+import numpy as np
+import collections
+from warnings import warn
+
+from nibabel.affines import apply_affine
+
+from .array_sequence import ArraySequence
+
+
+class UsageWarning(Warning):
+    pass
+
+
+class TractogramItem(object):
+    """ Class containing information about one streamline.
+
+    :class:`TractogramItem` objects have three main properties: `streamline`,
+    `data_for_streamline`, and `data_for_points`.
+
+    Parameters
+    ----------
+    streamline : ndarray shape (N, 3)
+        Points of this streamline represented as an ndarray of shape (N, 3)
+        where N is the number of points.
+    data_for_streamline : dict
+        Dictionary containing some data associated to this particular
+        streamline. Each key `k` is mapped to a ndarray of shape (Pk,), where
+        `Pt` is the dimension of the data associated with key `k`.
+    data_for_points : dict
+        Dictionary containing some data associated to each point of this
+        particular streamline. Each key `k` is mapped to a ndarray of
+        shape (Nt, Mk), where `Nt` is the number of points of this streamline
+        and `Mk` is the dimension of the data associated with key `k`.
+    """
+    def __init__(self, streamline, data_for_streamline, data_for_points):
+        self.streamline = np.asarray(streamline)
+        self.data_for_streamline = data_for_streamline
+        self.data_for_points = data_for_points
+
+    def __iter__(self):
+        return iter(self.streamline)
+
+    def __len__(self):
+        return len(self.streamline)
+
+
+class Tractogram(object):
+    """ Class containing information about streamlines.
+
+    Tractogram objects have three main properties: `streamlines`,
+    `data_per_streamline` and `data_per_point`.
+
+    Attributes
+    ----------
+    affine_to_rasmm : ndarray shape (4, 4)
+        Affine that brings the streamlines back to *RAS+* and *mm* space
+        where coordinate (0,0,0) refers to the center of the voxel.
+    """
+    class DataDict(collections.MutableMapping):
```

Comment:
Sorry - I haven't thought this full through yet, so would be grateful for pushback if I've missed the point.

Do you really need the DataDict class and subclasses?   I mean, do you really care about protecting these dictionaries from insertion of incorrect values outside the tractogram object?

I ask, because it would make the code simpler (I think) to put the setitem checks in the property code for the tractogram, as in `@data_per_streamline.setter` etc.    From the getter `@data_per_streamline`, you just return the dict.

What do you think?